### PR TITLE
Apple clip distance

### PR DESCRIPTION
**Pull request by palver123**
Implement [APPLE_CLIP_DISTANCE](https://www.khronos.org/registry/OpenGL/extensions/APPLE/APPLE_clip_distance.txt) OpenGL extension in ANGLE.

The implementation is not 100%.
 - `glEnable`/`glDisable` do not generate an error anymore when called with `CLIP_DISTANCEi_APPLE`, but they have no effect either. This means that the user can use clip planes in GLSL regardless of `glIsEnabled(CLIP_DISTANCEi_APPLE)`.
 - It might interfere with other extensions that use Geometry Shaders or Tessellator. It might even break them, not tested.

Tested manually with 2 clip planes:
multiple_clipplanes

### CODE REVIEW

File: src/libANGLE/renderer/d3d/DynamicHLSL.cpp [hunk 1]
```diff
@@ -594,6 +613,30 @@ void DynamicHLSL::generateShaderLinkHLSL(const gl::Caps &caps,
...
+    {
+        const auto semanticIdx  = i / 4;
+        const auto clipPlaneIdx = usedClipDistances[i];
+        vertexGenerateOutput << "    output.gl_ClipDistance" << semanticIdx << '.';
+        switch (i % 4)
+        {
+            case 0:
+                vertexGenerateOutput << 'x';
+                break;
+            case 1:
```
Comment (veghbalazs) [inline src/libANGLE/renderer/d3d/DynamicHLSL.cpp]: nit: you could use `op[]` on this vector instead of the switch-case swizzle, (tho you'll need a `float1` in the single component case above).

File: src/libANGLE/State.cpp [hunk 2]
```diff
@@ -999,7 +1000,12 @@ void State::setEnableFeature(GLenum feature, bool enabled)
         case GL_COLOR_MATERIAL:
             mGLES1State.mColorMaterialEnabled = enabled;
             break;
-        case GL_CLIP_PLANE0:
+        case GL_CLIP_DISTANCE0_APPLE:
```
Comment (lowdesertpunk) [inline src/libANGLE/State.cpp:1003]: What about keeping the original case too? In case the values change both functionality stays working.

File: src/libANGLE/Context.cpp [hunk 3]
```diff
@@ -1870,7 +1870,10 @@ void Context::getIntegervImpl(GLenum pname, GLint *params)
             *params = mState.mCaps.maxLights;
             break;
         case GL_MAX_CLIP_PLANES:
```
Comment (lowdesertpunk) [inline src/libANGLE/Context.cpp:1872]: Why not `GL_MAX_CLIP_DISTANCES_APPLE`?

File: src/libANGLE/Context.cpp [hunk 4]
```diff
@@ -3568,6 +3568,11 @@ void Context::initCaps()
         mState.mCaps.maxSmoothLineWidth            = 1.0f;
     }
 
+    if (getClientType() == EGL_OPENGL_API || getClientVersion() >= Version(3, 0))
```
Comment (lowdesertpunk) [inline src/libANGLE/Context.cpp]: In theory it could be supported in 2.0 too

File: src/libANGLE/renderer/d3d/DynamicHLSL.cpp [hunk 5]
```diff
@@ -1400,6 +1445,21 @@ void BuiltinVaryingsD3D::updateBuiltins(gl::ShaderType shaderType,
         builtins->glPosition.enable(userSemantic, reservedSemanticIndex++);
     }
 
+    if (metadata.getRendererMajorShaderModel() >= 4 &&
+        (shaderType == gl::ShaderType::Vertex || shaderType == gl::ShaderType::Geometry))
```
Comment (lowdesertpunk) [inline src/libANGLE/renderer/d3d/DynamicHLSL.cpp]: Shouldn't geometry shaders be just excluded?

File: src/libANGLE/State.cpp (hunk 2 above)
Comment (palver123) [reply src/libANGLE/State.cpp:1003]: I kept both functionalities. However GL_CLIP_PLANE0 and GL_CLIP_DISTANCE0_APPLE share the same value (by design) so you cannot write separate case blocks in the switch statement for them.

**Comment by palver123** [comment]
Measurement results:

### Full HD, NVidia GTX 1660 Super, no MSAA
| Workspace | `discard` in PS | `SV_ClipDistance` in VS |
| ------------- | ------------- | ------------- | 
| BigEngine | **7.966** ms | 7.97 ms |
| FrontSuspension | **1.564** ms | 1.6076 ms |

### Same with MSAA
| Workspace | `discard` in PS | `SV_ClipDistance` in VS |
| ------------- | ------------- | ------------- | 
| BigEngine | 8.601 ms | **8.007** ms |
| FrontSuspension | 2.156 ms | **1.958** ms |

### 4K, NVidia GTX 1070 (with MSAA)
| Workspace | `discard` in PS | `SV_ClipDistance` in VS |
| ------------- | ------------- | ------------- | 
| BigEngine | 15.39  ms | **13.16** ms |
| FrontSuspension | 2.08 ms | **1.55** ms |

It is clear that `SV_ClipDistance` scales better with bigger resolution (as expected). Even on Full HD resolution the benefits of `SV_ClipDistance` is measureable if we switch on MSAA. On the Surface Go tablet our earlier measurements suggest that the difference would be even bigger

File: src/libANGLE/renderer/d3d/DynamicHLSL.cpp (hunk 5 above)
Comment (palver123) [reply src/libANGLE/renderer/d3d/DynamicHLSL.cpp]: Well I was hoping to keep compatibility with geometry shaders but probably that would require more work to finish. Since we are not going to fix this extension for goemetry shaders (not it is required by the original spec.) it might cleaner to exclude it yes.

File: src/libANGLE/Context.cpp (hunk 3 above)
Comment (palver123) [reply src/libANGLE/Context.cpp:1872]: `GL_MAX_CLIP_DISTANCES_APPLE` has the same value as `GL_MAX_CLIP_PLANES`. Therefore I could not create separate case blocks for them.

File: src/libANGLE/Context.cpp (hunk 4 above)
Comment (palver123) [reply src/libANGLE/Context.cpp]: Probably yes. Will change.

File: src/libANGLE/renderer/d3d/DynamicHLSL.cpp (hunk 1 above)
Comment (palver123) [reply src/libANGLE/renderer/d3d/DynamicHLSL.cpp]: Changed it

File: src/libANGLE/Context.cpp (hunk 3 above)
Comment (lowdesertpunk) [reply src/libANGLE/Context.cpp:1872]: Let's put a `static_assert` for that then, both for future-proofing and for documentation purposes.

File: src/libANGLE/State.cpp (hunk 2 above)
Comment (lowdesertpunk) [reply src/libANGLE/State.cpp:1003]: Same for this.

File: src/libANGLE/State.cpp [hunk 6]
```diff
@@ -1110,7 +1116,10 @@ bool State::getEnableFeature(GLenum feature) const
             return mGLES1State.mRescaleNormalEnabled;
         case GL_COLOR_MATERIAL:
             return mGLES1State.mColorMaterialEnabled;
-        case GL_CLIP_PLANE0:
+        case GL_CLIP_DISTANCE0_APPLE:
+            if (mClientVersion >= Version(3, 0))
```
Comment (lowdesertpunk) [inline src/libANGLE/State.cpp]: also 2.0

File: src/libANGLE/validationES2.cpp [hunk 7]
```diff
@@ -841,6 +841,9 @@ bool ValidCap(const Context *context, GLenum cap, bool queryOnly)
         case GL_ROBUST_RESOURCE_INITIALIZATION_ANGLE:
             return queryOnly && context->getExtensions().robustResourceInitialization;
 
+        case GL_CLIP_DISTANCE0_APPLE:
+            return context->getClientVersion() < Version(2, 0) ||
+                   context->getClientVersion() >= Version(3, 0) && context->getExtensions().clipDistance;
```
Comment (lowdesertpunk) [inline src/libANGLE/validationES2.cpp]: here as well

File: src/libANGLE/Context.cpp [hunk 8]
```diff
@@ -1870,7 +1870,10 @@ void Context::getIntegervImpl(GLenum pname, GLint *params)
             *params = mState.mCaps.maxLights;
             break;
         case GL_MAX_CLIP_PLANES:
-            *params = mState.mCaps.maxClipPlanes;
+            if (mState.getClientVersion() >= Version(3, 0))
```
Comment (lowdesertpunk) [inline src/libANGLE/Context.cpp]: 2.0

File: src/libANGLE/validationES2.cpp [hunk 9]
```diff
@@ -842,8 +842,18 @@ bool ValidCap(const Context *context, GLenum cap, bool queryOnly)
...
 
         case GL_CLIP_DISTANCE0_APPLE:
+        case GL_CLIP_DISTANCE1_APPLE:
+        case GL_CLIP_DISTANCE2_APPLE:
+        case GL_CLIP_DISTANCE3_APPLE:
+        case GL_CLIP_DISTANCE4_APPLE:
+        case GL_CLIP_DISTANCE5_APPLE:
             return context->getClientVersion() < Version(2, 0) ||
-                   context->getClientVersion() >= Version(3, 0) && context->getExtensions().clipDistance;
+                   context->getClientVersion() >= Version(3, 0) &&
```
Comment (lowdesertpunk) [inline src/libANGLE/validationES2.cpp]: 2.0

File: src/libANGLE/validationES2.cpp [hunk 10]
```diff
@@ -842,8 +842,18 @@ bool ValidCap(const Context *context, GLenum cap, bool queryOnly)
...
+                       context->getExtensions().clipDistance;
+        case GL_CLIP_DISTANCE6_APPLE:
+        case GL_CLIP_DISTANCE7_APPLE:
+            return context->getClientVersion() >= Version(3, 0) &&
```
Comment (lowdesertpunk) [inline src/libANGLE/validationES2.cpp]: 2.0

**Review by lowdesertpunk — APPROVED** [review]

**Comment by lowdesertpunk** [comment]
All those changes in the generated files are a bit concerning. We will probably also get quite some conflicts when trying to update the `shapr_master` branch from upstream.

What about creating a separate branch (`shapr_patch_1`) without the changes in the generated files? This branch could be re-applied after resetting `shapr_master` to a new stable / beta branch without having to resolve all the conflicts.

**Comment by palver123** [comment]
#### What if we have conflicts in generated files?
Easy: always choose the latest.

#### How important is keeping these files up-to-date in the git repo?
Apparently not very much. Even if you checkout the current `chromium/4045` and run _scripts/run_code_generation.py_ you will see plenty of changes (provided that you touch one of the input JSONs first, e.g. _src/compiler/translator/builtin_variables.json_). This is because they do not always commit the latest result of code generation. Why? Maybe because we forked from an unstable (work in progress) branch. `chromium/4045` is not even included in the beta channel of win10 according to [OmahaProxy](http://omahaproxy.appspot.com/).

#### What if we want to rebase `shapr_master`?
We
 - rebase `shapr_master` and ignore the conflicts in generated files (choose local or choose remote version, does not matter it is a single click).
 - deal with conflicts in non-generated files
 - run the python script and then
 - look for compile/runtime errors.

#### How would this work with `shapr_patch_1`?
We
 - reset `shapr_master` to a new stable/beta branch
 - merge `shapr_patch_1` into `shapr_master` and 
 - deal with conflicts in non-generated files
 - run the python script and
 - look for compile/runtime errors.

I see an administrative overhead here. Not to mention checking out `shapr_patch_1` requires you to run the python script before you can continue. Checking out `shapr_master` as it is requires no additional action.

Tl;dr: I think the generated files won't hinder rebasing for us.

**Comment by lowdesertpunk** [comment]
Since these are generated files, they can be re-generated any time, after resolving conflict in non-generated files, fine.

My initial comment concentrated too much on generated files, but my main concern was actually the traceability of our changes. As we keep merging upstream changes, we might quickly loose track of what we modified. But that's why we are using pull-requests and not directly committing into the any master branch. Which was the case with our OpenCascade fork. So as long as we keep using pull-requests we will be fine, and won't end up in the situation like before when we didn't know what and why was changed.
